content/articles/custom-transformer/index.md

```diff
@@ -0,0 +1,204 @@
+      our_zip_file.extractall(path=our_path)
+      our_zip_file.close()
+
+get_data()
+```
+
+The code is for downloading the data from the URL to not dwell on it.
+We first import the `os` module for interacting with the Operating System. After that, we import the `tarfile` module for accessing and manipulating tar files. Lastly, we import the `urllib` for using URL manipulation functions.
+ 
+We then set our paths appropriately. In the `get_data()` function, we make a directory for our data, retrieve it from the URL then extracting and storing it. So, in your working directory, you will notice a directory called *datasets* created. On opening it, you will get another directory called *housing* with a file named **housing.csv** in it. We will use this file.
+
+We call the function.
+
+We then have this code for loading the CSV file.
+
+```python
+import pandas as pd
+
+def load_our_data(our_path=OUR_PATH):
+    our_file_path = os.path.join(our_path, "housing.csv")
+    return pd.read_csv(our_file_path)
+
+our_dataset = load_our_data()
+
+```
+We first import the `pandas` library, which loads the CSV data from the specified path, `our_file_path`.
+
+You can view the data using these codes:
+
+```python
+our_dataset.head()
+```
+
+```python
+our_dataset.info()
+```
+
+#### Cleaning the data
+
+The cleaning operation we will do here is filling empty numeric attributes with their median values. We will use the `SimpleImputer`, an estimator, to do that.
+
+```python 
+from sklearn.impute import SimpleImputer
+
+imputer = SimpleImputer(strategy="median")
+
+our_dataset_num = our_dataset.drop("ocean_proximity", axis=1)
+imputer.fit(our_dataset_num)
+X = imputer.transform(our_dataset_num)
+our_dataset_numeric = pd.DataFrame(X, columns=our_dataset_num.columns)
+
+```
+
+We drop the *ocean_proximity* attribute because it's a text attribute that will handle in the next section.
+
+The result produced is an array, so we convert it to a DataFrame.
+
+#### Handling text and categorical attributes
+
+We cannot handle text and numerical attributes similarly. For example, we cannot compute the median of text.
+
+We will use a transformer for this called the `OrdinalEncoder`. `OrdinalEncoder` is chosen because it is more pipeline friendly.
+
+```python
+from sklearn.preprocessing import OrdinalEncoder
+
+our_text_cats = our_dataset[['ocean_proximity']]
+our_encoder = OrdinalEncoder()
+our_encoded_dataset = our_encoder.fit_transform(our_text_cats)
+
+```
+
+### Our transformer
+
+This is where we will create the custom transformer. We will be adding these three attributes:
+- Rooms per household
+- Population per household
+- Bedrooms per household
+
+```python
+
+import numpy as np
+from sklearn.base import BaseEstimator, TransformerMixin
+#initialising column numbers
+rooms,  bedrooms, population, household = 3,4,5,6
+
+class CustomTransformer(BaseEstimator, TransformerMixin):
+    def __init__(self, add_bedrooms_per_room = True):
```

Comment:
Explain this class. How did you start to approach the logic? How did you break it into different methods? Explain the thought process step by step.